### PR TITLE
Add more comments to explain the code to generate the search index

### LINKED ISSUE #90766: Add code comments into `src/librustdoc/html/render/cache.rs`
The generation of the search index isn't easy to grasp so code comments would definitely improve the situation. It was started in https://github.com/rust-lang/rust/pull/90727 but needs to be continued.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #90766.

I tried to put comments when the code wasn't easy to understand at first sight and added more documentation on the recursive function. Please tell me if I misused the terminology or if comments can be improved or added into other places.

r? @notriddle

### CODE REVIEW

File: src/librustdoc/html/render/cache.rs [hunk 1]
```diff
@@ -360,6 +366,7 @@ crate fn get_real_types<'tcx>(
             insert_ty(res, tcx, arg.clone(), ty_generics);
         }
     } else {
+        // This is not a type parameter so we can add it directly (and look for its trait bounds).
```
Comment (notriddle) [inline src/librustdoc/html/render/cache.rs:369]: Do things that aren’t type parameters have trait bounds, or are we talking about its own nested type parameters.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/cache.rs:369]: Oh damn, I was talking about its own nested type parameters. Good catch!

**Comment by GuillaumeGomez** [comment]
I extended the comment and tried to make it more clear. Please tell me if it's too much or still unclear.

**Review by notriddle — APPROVED** [review]

**Pull request merged**